### DISCUSSION TITLE
How to calculate cache miss percentage?

### DISCUSSION THREAD

**Discussion by faradawn** [CMU-SAFARI/Pythia#9] [General]
Hi,

I tried the Cassandra trace on 1 core and rolled up the statistics:

<img width="742" alt="Screenshot 2023-06-12 at 8 23 12 AM" src="https://github.com/CMU-SAFARI/Pythia/assets/73060648/c2459e38-f158-430e-a757-7609924a0acc">

1. May I ask how to calculate the LLC_total_miss percentage?
2. Why does "nopref", which I assume is "no prefetcher", has the lowest miss rate?
3. In the first five lines, why pythisa has the highest IPC (instructions per cycle), even though it has more cache misses than "nopref"?

Thanks in advance!

**Comment by rahulbera** [comment]
Hi @faradawn, 
1. For calculating LLC miss ratio, you can compute the ratio of Core_0_LLC_total_miss/Core_0_LLC_total_access
2. There are four types of memory requests in ChampSim: load, RFO, prefetch, and writebacks. Core_0_LLC_total_miss is the sum of all four types of memory requests. From the screenshot, Pythia increases the number of all memory requests for cassandra_phase0_core0 by 37%. But if you check the Core_0_LLC_load_miss, Pythia reduces it by almost 52%.  This basically says the prefetch requests generated by Pythia made the load request misses go down, but the overall miss rate (including the prefetch request miss rate) goes high. Since a load miss is more critical to overall performance, the performance improves in the end.
3. The previous answer should address this question.

Hope this helps.

**Reply by faradawn** [reply]
Hi @rahulbera,

I found total_access in the raw experimental output:
```
/Pythia/experiments/experiments_1C/cassandra_phase0_core0_nopref.out'
```
and calculated the miss rate for Nopref and Pythia
```
Nopref: 1056690 / 1784950 = 0.59
Pythia: 1448139 / {2522886 = 0.57
```

Got that Pythia used more memory access (prefetching) to reduce load cache miss, which is more crucial to performance (instructions per cycle). 

It makes sense now. 

Thanks for the clear explanation!
